### ISSUE TITLE
Render resolution is 512

### ISSUE THREAD

**Issue by Michaelwhite34** [abdallahdib/NextFace#31]
I notice that render and debug output all have resolution 512, does that mean per-pixel basis is based on this 512 image ? I just tried 2048 uv and feel that the output is blurry.

**Comment by Michaelwhite34** [comment]
The processing is not over yet (it takes around 12 hours on cpu), currently at step3. The render is still noisy even if I set render sample to 20000, does that mean I need to use a bigger number? Or I need to change another parameter?
![debug2_iter60_frame0](https://user-images.githubusercontent.com/72028833/181397519-4dcb71f2-a27e-4ba4-a9e8-5efc71d91a21.jpg)

**Comment by Michaelwhite34** [comment]
When I import the diffence map into photoshop and apply mean filter, the mean loss won't drop from step 3 30.

**Comment by abdallahdib** [comment]
hi 
the debug renderer is always noisy as during optimization we use only  8 samples per pixel (for faster optimization)
after the optimization is finished the final image is rendered at high number of samples 
please refer to these parameters in optimConfig.ini: 

rtSamples = 4000 #the number of ray tracer samples to render the final output (higher is better but slower) best value is  20000  but on my old gpu it takes too much time to render. if u have nvidia rtx u are fine enjoy :) 
rtTrainingSamples = 8#number of ray tracing to use during training

**Comment by Michaelwhite34** [comment]
> hi the debug renderer is always noisy as during optimization we use only 8 samples per pixel (for faster optimization) after the optimization is finished the final image is rendered at high number of samples please refer to these parameters in optimConfig.ini:
> 
> rtSamples = 4000 #the number of ray tracer samples to render the final output (higher is better but slower) best value is 20000 but on my old gpu it takes too much time to render. if u have nvidia rtx u are fine enjoy :) rtTrainingSamples = 8#number of ray tracing to use during training

Yeah, after a day I figured that out, I think that default 8 sample is too low, I used 100 sample the result is great.But for now I have difficulty reproducing the same render result in blender.

**Comment by Michaelwhite34** [comment]
I took a screen shot of your video
![landmarks0](https://user-images.githubusercontent.com/72028833/181696133-64ce66b9-f106-41b1-891c-1c726043bd2c.png)
![debug2_iter0_frame0](https://user-images.githubusercontent.com/72028833/181696200-24bf8bfc-a82f-410a-ada1-e3b36d95617d.png)
![envMap_0](https://user-images.githubusercontent.com/72028833/181696221-74812833-7879-43cd-8c9c-46a66e268830.png)
The color doesn't even match.

**Comment by Michaelwhite34** [comment]
Ok, I think I have found the rotation, but there is still difference in the render Here is the comparsion
![render](https://user-images.githubusercontent.com/72028833/181711150-ce3b337b-3b12-4004-89ef-a0a075d34a10.jpg)
. blender filmic
![render_standrad_color](https://user-images.githubusercontent.com/72028833/181711383-c36ec3f2-b960-4c65-886b-06b824f9b0db.jpg)
blender standard
![debug2_iter0_frame0](https://user-images.githubusercontent.com/72028833/181711706-6a751be8-fde1-4fa9-849a-c1c876e68a5d.jpg)
debug render
I just use principle bsdf with direct diffuse roughness and specular input.Environment map has strength 1.

**Comment by abdallahdib** [comment]
Hi Michael, the estimated maps (diffuse specular albedo) should be compatible with rendering engines (such as blender). 
I am wondering whatever blender filmic or standard adds any post processing to the final output. Also did u verify if any gamma correction is needed for blender? it seems that blender standard is closer to the output of next face than filmic. 


Also I would like to know how did u use the environment map in blender to render the final mesh. There is a github issue who tried to to the same here (https://github.com/abdallahdib/NextFace/issues/6)
can u please share more information on this (a video or explanation on how to do it would be great so that i can add it to the readme of the library)

**Comment by Michaelwhite34** [comment]
> Hi Michael, the estimated maps (diffuse specular albedo) should be compatible with rendering engines (such as blender). I am wondering whatever blender filmic or standard adds any post processing to the final output. Also did u verify if any gamma correction is needed for blender? it seems that blender standard is closer to the output of next face than filmic.
> 
> Also I would like to know how did u use the environment map in blender to render the final mesh. There is a github issue who tried to to the same here (#6) can u please share more information on this (a video or explanation on how to do it would be great so that i can add it to the readme of the library)

It might be the environment resolution. I just use a 4096 environment map on another example ,it looks much better.Please add option to directly export specific map from pickle file, and add time estimation for that, sometimes I really have no idea it will take hours, which makes a lot anxiety.

**Comment by abdallahdib** [comment]
the environment maps is already exported as an exr or png file (refer to the config file) 
which environment map are did u use for blender ?

**Comment by Michaelwhite34** [comment]
> the environment maps is already exported as an exr or png file (refer to the config file) which environment map are did u use for blender ?

I mean I change the export environment resolution setting to 4096, it is 64 by default.

**Comment by abdallahdib** [comment]
well if u change to 4096 this resolution will be used for the env map during optimization as well. not sure if u need that resolution. what did u get with the 64x64 resolution with blender? 

yes indeed the focal length is inside the pickle file

**Comment by Michaelwhite34** [comment]
> well if u change to 4096 this resolution will be used for the env map during optimization as well. not sure if u need that resolution. what did u get with the 64x64 resolution with blender?
> 
> yes indeed the focal length is inside the pickle file
Yeah, but it will be easier to read if you print that in teriminal,at the beginning for example.

**Comment by Michaelwhite34** [comment]
> well if u change to 4096 this resolution will be used for the env map during optimization as well. not sure if u need that resolution. what did u get with the 64x64 resolution with blender?
> 
> yes indeed the focal length is inside the pickle file

Yeah, but it will be easier to read if you print that in teriminal,at the beginning for example.

**Comment by abdallahdib** [comment]
will add an export of the estimated camera parameters into a text file.

**Comment by Michaelwhite34** [comment]
> will add an export of the estimated camera parameters into a text file.

And this ... add option to directly export specific map from pickle file, and add time estimation for that
and maybe add option to resume optimization at certain stage (2 or 3)

**Comment by abdallahdib** [comment]
textures are already saved in the output directory. 
u can already resume optimization from a given pickle file (use the flag --checkpoint) and u can skip step 1 and/or 2 and/or 3
by adding these flags: --skipStage1 --skipStage2 --skipStage3

**Comment by abdallahdib** [comment]
plz refer to the main function of optimizer.py

**Comment by Michaelwhite34** [comment]
> textures are already saved in the output directory. u can already resume optimization from a given pickle file (use the flag --checkpoint) and u can skip step 1 and/or 2 and/or 3 by adding these flags: --skipStage1 --skipStage2 --skipStage3

Sometimes we want to use different setting for texture, so we can just use --usecheckpoint --skipStage1 --skipStage2 --skipStage3 at the same time to start from the export process ? But we may only want to export specific maps only.

**Comment by abdallahdib** [comment]
if u use --usecheckpoint --skipStage1 --skipStage2 --skipStage3, it will only load the pickle file and save the output (no optimization is done then) i m not sure if this is useful. U an always customize the code to fit ur specific needs.

**Comment by Michaelwhite34** [comment]
> if u use --usecheckpoint --skipStage1 --skipStage2 --skipStage3, it will only load the pickle file and save the output (no optimization is done then) i m not sure if this is useful. U an always customize the code to fit ur specific needs.

Yeah, that will be nice. Will you add option to export specific map only ?, for me I can read some code but honestly it's difficult for me to modify for my usage.

**Comment by Michaelwhite34** [comment]
> Hi Michael, the estimated maps (diffuse specular albedo) should be compatible with rendering engines (such as blender). I am wondering whatever blender filmic or standard adds any post processing to the final output. Also did u verify if any gamma correction is needed for blender? it seems that blender standard is closer to the output of next face than filmic.
> 
> Also I would like to know how did u use the environment map in blender to render the final mesh. There is a github issue who tried to to the same here (#6) can u please share more information on this (a video or explanation on how to do it would be great so that i can add it to the readme of the library)

For obj import with forward -Z, up Y.  Rotation X 90, Y 180. Camera rotation X -90, Y 180. For environment map, rotate along Z 90, rotate along X  180, that should work.

**Comment by Michaelwhite34** [comment]
> if u use --usecheckpoint --skipStage1 --skipStage2 --skipStage3, it will only load the pickle file and save the output (no optimization is done then) i m not sure if this is useful. U an always customize the code to fit ur specific needs.

Just tried it, it outputs nothing in this way after waiting for a long time.

**Comment by Michaelwhite34** [comment]
When I tried to export 8192 environment, it says cpu out of memory. I have 64G memory though.

**Comment by Michaelwhite34** [comment]
> Hi Michael, the estimated maps (diffuse specular albedo) should be compatible with rendering engines (such as blender). I am wondering whatever blender filmic or standard adds any post processing to the final output. Also did u verify if any gamma correction is needed for blender? it seems that blender standard is closer to the output of next face than filmic.
> 
> Also I would like to know how did u use the environment map in blender to render the final mesh. There is a github issue who tried to to the same here (#6) can u please share more information on this (a video or explanation on how to do it would be great so that i can add it to the readme of the library)
Can you render in redner without device color transform and color space conversion (raw)? That way we can actually compare the difference render between blender and redner

**Comment by Michaelwhite34** [comment]
redner uses phong reflection, which is outdated for modern render engines, it might be the reason that we can't get similar result.

**Comment by Michaelwhite34** [comment]
![render__11](https://user-images.githubusercontent.com/72028833/182012314-b4ede7c2-c440-4b43-8c71-8ad86c8e48b8.png)
debug light
![debug light](https://user-images.githubusercontent.com/72028833/182012293-0c1082a4-1db4-447d-ae34-36cee135e339.png)
blender light  filmic

**Comment by Michaelwhite34** [comment]
By the way is the exported env map expected to be directly used as environment color map? As I see the debug render doesn’t have visible colored reflection.

**Comment by Refl-ex** [comment]
> I just use principle bsdf with direct diffuse roughness and specular input.Environment map has strength 1.
Can you share the blender node structure? My result is so different from origin using direct diffuse roughness and specular input to principle bsdf.
